core, core/vm, tests: changed the initialisation behaviour of the EVM

obscuren opened the pull request:
### Refactor EVM initialisation 

* The EVM was previously initialised and created for every `CALL`, `CALLCODE`,
`DELEGATECALL` and `CREATE`. This PR changes this behaviour so that the same
EVM can be used through the session and beyond as long as the
Environment sticks around.
* Renamed `Vm` to `EVM`
* Additionally this removes the global EVM jump table and initialises a new jump table for each new instance of the EVM, removes the global `Debug` variable from package `core/vm` and includes a new `vm.Config` option which is passed to the `EVM` initialiser.
* Changed structured logging to include the following configuration options:
    * Disable storage capture
    * Disable memory capture
    * Disable stack capture
    * Switch between full storage capture or changed values capture 

### Debugging options

Refactored `eth/api` debug slightly:

* Renamed `debug.replayTransaction` to `debug.traceTransaction`
* Changed `debug.traceTransaction` and removed stack, storage & memory capture size and added a logger option object (see above):
```
{disableStorage: bool, disableMemory: bool, disableStack: bool, fullStorage: bool}
```
* Adds the following block debugging methods to `debug`
    *  `traceBlock(rlp)`
    * `traceBlockFromFile(fileName)`
    * `traceBlockFromNumber(number)`
    * `traceBlockFromHash(hash)`
    * `traceCall(callopts)`

### `debug.traceBlockXXX`

The debug block trace methods execute the given block without apply the state to disk and return you the results of the execution of the block, including the log traces (structured logs) generated by the transaction invocations.

### `debug.traceCall`

Trace call is similar to the regular `eth.call` method, but in addition to returning the execution results, gas and/or an error it also returns you the full log trace (structured logs) generated by the call.

robotally commented:
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Thu Jan 21 14:31:36 UTC 2016_

obscuren commented:
@ppratscher want to give you a heads up of these changes. I know you're using `debug.replayTransaction`. This will be renamed to just plainly `debug.transaction`.

ppratscher commented:
Thanks for the update. Once this PR is merged I will update to the new method signature.

obscuren commented:
@fjl @Gustav-Simonsson @karalabe Can I get some early review on this before this turns in to a huge PR.

karalabe commented on cmd/evm/main.go:
```diff
@@ -174,17 +176,23 @@ type VMEnv struct {
-func NewEnv(state *state.StateDB, transactor common.Address, value *big.Int) *VMEnv {
-	return &VMEnv{
+func NewEnv(state *state.StateDB, transactor common.Address, value *big.Int, cfg *vm.Config) *VMEnv {
+	params.HomesteadBlock = new(big.Int)
```
Is it wise to reset the HomesteadBlock here? Any code calling NewEnv post startup would immediately blow up consensus.

obscuren replied:
This is the EVM intepreter utility binary. I don't think that's going to matter. In any case the number is settable in my follow up PR.

karalabe replied:
Ah indeed, didn't see that it's not Geth code, sorry.

karalabe commented on cmd/utils/flags.go:
```diff
@@ -203,11 +203,6 @@ var (
-	VMDebugFlag = cli.BoolFlag{
-		Name:  "vmdebug",
-		Usage: "Virtual Machine debug output",
-	}
```
What was that comment again? :trollface:

obscuren replied:
You are faggot

(yes exactly like that)

karalabe commented on core/execution.go:
```diff
@@ -60,7 +60,8 @@ func Create(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPric
 
 func exec(env vm.Environment, caller vm.ContractRef, address, codeAddr *common.Address, input, code []byte, gas, gasPrice, value *big.Int) (ret []byte, addr common.Address, err error) {
-	evm := vm.NewVm(env)
+	//evm := vm.NewVm(env)
```
I guess this and its counterpart in the next diff segment can be dropped?

karalabe commented on core/state/state_object.go:
```diff
@@ -126,31 +131,32 @@ func (self *StateObject) Storage() Storage {
 
 func (self *StateObject) SetState(k, value common.Hash) {
-	self.storage[k.Str()] = value
+	self.storage[k] = storageValue{value: value, dirty: true}
```
I know it's not your code, but can we rename `k` to `key`?

obscuren replied:
Sure thing. I'll change it

karalabe commented on core/state/state_object.go:
```diff
@@ -126,31 +131,32 @@ func (self *StateObject) Storage() Storage {
+				self.trie.Delete(key[:])
+				continue
+			}
+			self.setAddr(key, sv.value)
```
Shouldn't the dirty flag be set to false here? Not sure if an Update can be rolled back/discarded, but if not, then I guess there's no point in updating multiple times.

obscuren replied:
No, updates don't get rolled back.

obscuren replied:
Also storage values are stack allocated and not heap so that wouldn't work without a change to the storage mechanism itself.

obscuren replied:
As discussed on gitter this will stay the way it is.Fixing this proper is going to require significant changes in the core which may lead to consensus failures. This shouldn't be changed for 1.4

karalabe commented on core/state/state_object.go:
```diff
@@ -245,10 +251,34 @@ func (self *StateObject) Value() *big.Int {
+			cb(key, common.BytesToHash(it.Value))
+		}
+	}
+}
```
Hmm, this method will return duplicates if full is requested for both modified as well as requested keys. Shouldn't there rather be two separate sections?

```go
if !full {
	for h, sv := range self.storage {
 		if sv.dirty {
 			cb(h, sv.value)
 		}
	}
	return
}
it := self.trie.Iterator()
	for it.Next() {
		// ignore cached values
		key := common.BytesToHash(self.trie.GetKey(it.Key))
		if _, ok := self.storage[key]; !ok {
			cb(key, common.BytesToHash(it.Value))
		}
	}
```

karalabe replied:
Question: since this method depends on the dirty flag, we may not be able to set it to false above. Maybe we could maintain two flags instead? `dirty` to be used by the state object to avoid too many database writes and/or trie ops, and `modified` to track whether we should "diff" it or not?

obscuren replied:
Yeah, I think that last approach should be quite fine. Regarding the former, you're right, it will return dups. Thanks

obscuren replied:
Looking at your example again it will actual display incorrect values (i.e. values that are old, this is what the cache is for).

obscuren replied:
As per gitter. My code is fine :-P

karalabe replied:
:+1:

karalabe commented on core/state/state_object.go:
```diff
@@ -261,6 +291,17 @@ func (self *StateObject) EachStorage(cb func(key, value []byte)) {
+		}
+	}
+}
+*/
```
I think these can be deleted from above.

obscuren replied:
:+1:

karalabe commented on core/vm/environment.go:
```diff
@@ -76,6 +74,10 @@ type Environment interface {
 	Create(me ContractRef, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error)
 }
 
+type Vm interface {
```
Please add a doc :P

obscuren replied:
Doh, thanks

karalabe commented on core/vm/environment.go:
```diff
@@ -76,6 +74,10 @@ type Environment interface {
 }
 
+type Vm interface {
+	Run(*Contract, []byte) ([]byte, error)
```
Here too :P

karalabe commented on core/vm/vm.go:
```diff
@@ -363,24 +385,34 @@ func (self *Vm) RunPrecompiled(p *PrecompiledAccount, input []byte, contract *Co
-		for i, item := range stack.Data() {
-			stck[i] = new(big.Int).Set(item)
+func (evm *EVM) log(pc uint64, op OpCode, gas, cost *big.Int, memory *Memory, stack *stack, contract *Contract, err error) {
+	if evm.cfg.Debug {
```
It's cleaner to do a 

```
if !evm.cfg.Debug {
  return
}
// all other code here
```

obscuren replied:
True. Not sure why I didn't do that. I usually use that construct.

karalabe commented on core/vm_env.go:
```diff
@@ -41,29 +41,38 @@ func GetHashFn(ref common.Hash, chain *BlockChain) func(n uint64) common.Hash {
-
-	getHashFn func(uint64) common.Hash
-	// structured logging
+	// Message appliod
```
Could you move the comments after the fields? Looks weird this way :P Also `appliod`

obscuren replied:
Appliod is a thing you know.

obscuren replied:
BTW, I'm not sure about the commenting *after* the fields. Why? Is there a non-written  rule that interfaces get document **above** and structs fields after?

karalabe replied:
No, just that generally writing comments after fields makes the code more condensed and easier to skim through. It also aligns the comments nicely.

Usually with interfaces the method name is already long enough so adding comment after it just makes it unreadable and also method comments tend to be much longer, usually multiple lines whereas field comments are only a few words.

obscuren replied:
Fair enough

karalabe commented on eth/api.go:
```diff
@@ -1506,17 +1508,58 @@ func NewPrivateDebugAPI(eth *Ethereum) *PrivateDebugAPI {
+	blockRlp, err := ioutil.ReadAll(file)
+	if err != nil {
+		return false, fmt.Errorf("could not read file: %v", err)
+	}
```
[`ioutil.ReadFile`](https://golang.org/pkg/io/ioutil/#ReadFile) :) Less code and cleaner

obscuren replied:
I always forget about that function. Reminds of a discussion @fjl and I had a long time ago.

karalabe commented on eth/api.go:
```diff
@@ -1506,17 +1508,58 @@ func NewPrivateDebugAPI(eth *Ethereum) *PrivateDebugAPI {
-func (api *PrivateDebugAPI) ProcessBlock(number uint64) (bool, error) {
+// Block processes the given block's RLP but does not import the block in to
+// the chain.
+func (api *PrivateDebugAPI) Block(blockRlp []byte, logger VmLoggerOptions) (bool, error) {
```
I'm not sure it's a good idea to drop the "verb" from the method name. Although I get that `debug.Block` looks fancy, imho it's not clear that debug is namespace and verb at the same time. I'd vote to add back either the process or something `debug.ProcessBlock` `debug.RunBlock` `debug.DebugBlock`. Not sure what is the best, but simple debug.Block doesn't really tell what it does.

karalabe replied:
`ReplayBlock` perhaps

obscuren replied:
Agree, though `ProcessBlock` is quite a shitty name. `debug.TraceBlock` is something maybe

karalabe replied:
I can go with Trance, nice and to the point.

obscuren replied:
Oomph-ta oomph-ta oomph-ta-ta-ta oomph oomph oomph oomph

karalabe commented on eth/api.go:
```diff
@@ -1545,7 +1591,16 @@ func (api *PrivateDebugAPI) SetHead(number uint64) {
 }
 
-// StructLogRes stores a structured log emitted by the EVM while replaying a
+// ExecutionRes groups all structured logs emitted by the EVM
```
Name in the doc not the same as the type.

obscuren replied:
Thanks

trapp commented:
Please consider adding a `depth` value to the output of each log in debug.transaction with this refactoring step: obscuren/go-ethereum/pull/5

> It adds the depth to each log output which makes it easier to track the status of CALL opcodes that call other contracts. The next opcode starts again with pc: 0 which makes it obvious that execution went down one level but tracking when it goes back up is not so trivial.

> Adding the depth value to each log makes it easier.

fjl commented on core/vm/vm.go:
```diff
@@ -28,27 +28,49 @@ import (
+	DisableStack   bool // disable stack capture
+	DisableStorage bool // disable storage capture
+	FullStorage    bool // show full storage (slow)
 }
```
Can the log options be in Config? They'd need slightly different names maybe, but I don't see why they're separate.

fjl replied:
You could call them LogFullStorage, NoMemoryLog, NoStackLog...

fjl commented on eth/api.go:
```diff
@@ -1558,16 +1613,73 @@ type structLogRes struct {
+
+// Transaction returns the structured logs created duritg the execution of EVM
+// and returns them as a JSON object.
+func (s *PrivateDebugAPI) Transaction(txHash common.Hash, logger VmLoggerOptions) (*ExecutionResult, error) {
```
The type of logger can be vm.LogCfg, then you don't need to convert it later.

obscuren replied:
Thanks, I'll change that. It's a left over from the previous PR

obscuren commented:
I'll rename the current tracing methods to `debug.traceX`, including translations.

Additionally I'll add an `debug.traceCall` performing the same action as `eth.call`, however returning the full trace of the call.

Current block trace methods do not yet return the trace in JavaScript. I'll need to address that, though for the rest of the changes in this PR it shouldn't matter.

obscuren commented:
@trapp really cool you made a PR on this PR, thanks! I'll merge that in to this PR

fjl commented on core/state_transition.go:
```diff
@@ -253,7 +253,8 @@ func (self *StateTransition) transitionDb() (ret []byte, usedGas *big.Int, err e
 
-	if vm.Debug {
+	logs := vmenv.StructLogs()
+	if len(logs) > 0 {
```
This change makes it print all struct logs to stderr. Why did you add this?
If users (e.g. etherchain) request struct logs through the API, their logs grow very large.

obscuren replied:
In order to have structured logs created you need to specify it in the config. If that's not set no structured logs will ever be emitted.

This flag has been completely removed. I'm pondering to remove this bit of code completely in favour of the new structured logging (e.g. json). Have another tool convert that to the structured log output generated by the `vm.StdErrFormat`.

fjl replied:
Yes, I think it should go away. We still need `StdErrFormat` for cmd/evm, but it's fine to have it. Maybe it should be  `PrintStructLogs(io.Writer)` but that can also be done later.

obscuren commented:
I've refactored the EVM logging in to a separate object. Refactored how log traces are emitted (no longer through the `vm.Environment`. Added a `vm.Collector` interface to `vm.LogCfg` which allows external tools to set for collecting logs. Additionally I've removed the default logging from the `core.StateTransaction` model.

trapp commented:
I updated the depth extension to match your latest changes: obscuren/go-ethereum/pull/6

obscuren commented:
@trapp I was just about to ping you! Nice work!

obscuren commented:
@ppratscher another headsup, the method has been renamed to `debug.traceTransaction`.

obscuren commented:
Structured logs memory formatting has changed to:

```
{
      error: null,
      gas: 100000,
      gasCost: 0,
      memory: ["0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000000", "0000000000000000000000000000000000000000000000000000000000000060", "000000000000000000000000a0f5578094e713c58b901ec43fb58bbbd7a63372", "0000000000000000000000000000000000000000000000004563918244f40000"],
      op: "STOP",
      pc: 351,
      stack: [],
      storage: {}
  }
```

Storage remains the same

karalabe commented on core/blockchain.go:
```diff
@@ -162,6 +163,10 @@ func NewBlockChain(chainDb ethdb.Database, pow pow.PoW, mux *event.TypeMux) (*Bl
 	return bc, nil
 }
 
+func (self *BlockChain) SetConfig(vmConfig *vm.Config) {
```
Doc :P

karalabe replied:
Also please mention that this method is not thread safe. I.e. after you start using the blockchain, you cannot call this method any more.

obscuren replied:
Don't worry about this bit. It's removed in https://github.com/ethereum/go-ethereum/pull/2281 I don't expect the two to take very long to get merged after eachother. It's just a quick workaround to enable jit. We could drop this commit all together if that makes more sense.

karalabe commented on core/vm/logger.go:
```diff
@@ -18,12 +18,154 @@ package vm
+
+// VmLogger is a logger interface used by the EVM to emit structured logs.
+type VmLogger interface {
+	// Log logs the given input.
```
Method doc doesn't conform to func name `CaptureState`.

obscuren replied:
Ah thanks. I renamed the method but didn't update the method doc.

karalabe commented on core/vm/logger.go:
```diff
@@ -18,12 +18,154 @@ package vm
+	Err     error
+}
+
+// logger implements VmLogger
```
Capital L

karalabe commented on core/vm/logger.go:
```diff
@@ -18,12 +18,154 @@ package vm
+	}
+
+	// Copy the storage based on the settings specified in the log config. If full storage
+	// is disabled (default) we can use the simple Stocage.Copy method, otherwise we use
```
Small typo `Stocage.Copy`.

karalabe commented on core/vm/logger.go:
```diff
@@ -18,12 +18,154 @@ package vm
+	// is disabled (default) we can use the simple Stocage.Copy method, otherwise we use
+	// the state object to query for all values (slow process).
+	var storage Storage
+	fmt.Println(l.cfg.DisableStorage, l.cfg.FullStorage)
```
I think you didn't want this `Println` in here :P

obscuren replied:
Oops :)

karalabe commented on core/vm/logger.go:
```diff
@@ -61,7 +203,7 @@ func StdErrFormat(logs []StructLog) {
 		fmt.Fprintln(os.Stderr, "STORAGE =", len(log.Storage))
 		for h, item := range log.Storage {
-			fmt.Fprintf(os.Stderr, "%x: %x\n", h, common.LeftPadBytes(item, 32))
+			fmt.Fprintf(os.Stderr, "%x: %x\n", h, item, 32)
```
This change half correct, as it adds an extra 32 parameter to the end of the `Fprintf`.

obscuren replied:
Hmm, this must be an accident. The change doesn't make sense at all.

karalabe commented on core/vm/logger_test.go:
```diff
@@ -0,0 +1,87 @@
+package vm
```
Copyright note :)

obscuren replied:
Cheers

karalabe commented on core/vm/vm.go:
```diff
@@ -28,27 +28,52 @@ import (
-func New(env Environment) *Vm {
-	// init the jump table. Also prepares the homestead changes
-	jumpTable.init(env.BlockNumber())
+func New(env Environment, cfg *Config) *EVM {
```
Please add a doc here too.

karalabe commented on core/vm/vm.go:
```diff
@@ -28,27 +28,52 @@ import (
+	Logger    LogCfg
+}
+
 // Vm is an EVM and implements VirtualMachine
```
Type name was updated, please update the doc accordingly.

obscuren commented:
@karalabe comments addressed PTAL

karalabe commented on eth/api.go:
```diff
@@ -1506,38 +1508,111 @@ func NewPrivateDebugAPI(eth *Ethereum) *PrivateDebugAPI {
 
-// ProcessBlock reprocesses an already owned block.
-func (api *PrivateDebugAPI) ProcessBlock(number uint64) (bool, error) {
+type BlockTraceResult struct {
```
Doc :P

karalabe commented on eth/api.go:
```diff
@@ -1506,38 +1508,111 @@ func NewPrivateDebugAPI(eth *Ethereum) *PrivateDebugAPI {
+	Error      error          `json:"error"`
+}
+
+// Block processes the given block's RLP but does not import the block in to
```
Please fix method docs to include `Trace`

karalabe commented on eth/api.go:
```diff
@@ -1545,51 +1620,104 @@ func (api *PrivateDebugAPI) SetHead(number uint64) {
+type ExecutionResult struct {
+	Gas         *big.Int       `json:"gas"`
+	ReturnValue string         `json:"returnValue"`
+	StructLogs  []structLogRes `json:"structLogs"`
```
Can we also return whether the transaction failed or succeeded? (i.e. throw or out of gas?)

obscuren replied:
No need. All you gotta do is `structLogRes[len(structLogRes)-1].err == nil`

karalabe replied:
Ah, indeed, I didn't see that the OP has an error field.

karalabe commented on eth/api.go:
```diff
@@ -1545,51 +1620,104 @@ func (api *PrivateDebugAPI) SetHead(number uint64) {
+		addr := 0
+		for i := 0; i+32 <= len(trace.Memory); i += 32 {
+			formattedStructLogs[index].Memory = append(formattedStructLogs[index].Memory, fmt.Sprintf("%x", trace.Memory[i:i+32]))
+			addr++
```
Is `addr` used for anything?

obscuren replied:
No, not anymore. This was used before when the `formatterStructLogs[].Memory` was a `map[string]string`. Will remove.

karalabe commented on eth/api.go:
```diff
@@ -1545,51 +1620,104 @@ func (api *PrivateDebugAPI) SetHead(number uint64) {
+	return formattedStructLogs
+}
+
+// TraceTransaction returns the structured logs created duritg the execution of EVM
```
`duritg` ;)

karalabe commented on eth/api.go:
```diff
@@ -1601,87 +1729,76 @@ func (s *PrivateDebugAPI) doReplayTransaction(txHash common.Hash) ([]vm.StructLo
 	ret, gas, err := core.ApplyMessage(vmenv, msg, gp)
 	if err != nil {
-		return nil, nil, nil, fmt.Errorf("Error executing transaction %v", err)
+		return nil, fmt.Errorf("Error executing transaction %v", err)
```
If a transaction faults or runs out of gas, is that an error returned by apply message? Just making sure I understand correctly.

obscuren replied:
No. Out of gas is not an error. This would actually return a consensus error if it were to be included in a block:

* not enough funds
* not enough intrinsic gas
* invalid `v`, `r`, `s` values

karalabe replied:
Ok, that is what I thought, just figured I'd double check. Thanks for the explanation!

karalabe commented on eth/api.go:
```diff
@@ -1601,87 +1729,76 @@ func (s *PrivateDebugAPI) doReplayTransaction(txHash common.Hash) ([]vm.StructLo
-			i++
-		}
+	ret, gas, err := core.ApplyMessage(vmenv, msg, gp)
+	if len(ret) == 0 { // backwards compatibility
```
Backward compatibility with what?

obscuren replied:
*shrug*... copied. I'm going to remove that check all together. I've no idea about the intention.

karalabe replied:
@bas-vk Any ideas of what this "backwards compatibility" entailed?

karalabe commented on eth/api.go:
```diff
@@ -1506,38 +1508,113 @@ func NewPrivateDebugAPI(eth *Ethereum) *PrivateDebugAPI {
+	}
+}
+
+// BlockFromFile loads the block's RLP from the given file name and attempts to
```
My previous comment was fix the missing Traces, as in plural :P

obscuren replied:
:goberserk:

karalabe commented on eth/api.go:
```diff
@@ -1506,38 +1508,113 @@ func NewPrivateDebugAPI(eth *Ethereum) *PrivateDebugAPI {
+	return api.TraceBlock(blockRlp, config)
+}
+
+// ProcessBlock processes the block by canonical block number.
```
Here too

karalabe commented on eth/api.go:
```diff
@@ -1506,38 +1508,113 @@ func NewPrivateDebugAPI(eth *Ethereum) *PrivateDebugAPI {
+	}
+}
+
+// BlockByHash processes the block by hash.
```
Here too

karalabe commented:
LGTM :+1:

frozeman commented:
This is probably interesting in web3.js, but the debug is not part of official web3.js yet, so it can be added in the go console by @bas-vk. It would be nice to have an idea how the outputs look like.

fjl commented on core/vm/logger.go:
```diff
@@ -18,12 +18,157 @@ package vm
+type VmLogger interface {
+	// CaptureState caputeres the current EVM state and logs it.
+	CaptureState(pc uint64, op OpCode, gas, cost *big.Int, memory *Memory, stack *stack, contract *Contract, err error)
+}
```
This interface is pointless. Users cannot supply their own implementation and they probably wouldn't want to (they'd just be reimplementing Logger).

I suggest that you 

* remove `VmLogger`

* unexport `Logger`

* make type of the `EVM.logger` field `*logger` and leave it nil unless `Config.Debug` is true

* change `EVM.Run` so it does

  ```go
  if evm.cfg.Debug {
      evm.logger.CaptureState(...)
  }
  ```
  
It will be less code overall.

obscuren replied:
You are right. This was actually case when I had a nop logger in this PR. I'll remove it as you suggested.

fjl commented on core/vm/logger.go:
```diff
@@ -18,12 +18,157 @@ package vm
+type StructuredLogCollector interface {
+	// Adds the structured log to the collector.
+	AddStructuredLog(StructLog)
+}
```
I do like the fact that log collection is an interface.

The naming is weird though. I think you should go for either `StructuredLog` or `StructLog`
and be consistent about it. I'd personally call the method `StructLog` and the type `StructLogger` because
they are shorter. Note that no interface ambiguity can arise because the parameter is `vm.StructLog`.

fjl commented on core/vm/logger.go:
```diff
@@ -18,12 +18,157 @@ package vm
+	DisableStorage bool                   // disable storage capture
+	FullStorage    bool                   // show full storage (slow)
+	Collector      StructuredLogCollector // the log collector
+}
```
More naming nits (if only there was something else I could complain about...):
Why is Cfg abbreviated here but Config isn't?

obscuren replied:
No particular reason comes to mind. Just poor naming I guess.

fjl commented on eth/api.go:
```diff
@@ -1545,51 +1622,102 @@ func (api *PrivateDebugAPI) SetHead(number uint64) {
+		DisableStorage: opts.DisableStorage,
+		FullStorage:    opts.FullStorage,
+	}
+}
```
Why is this still here?

obscuren replied:
Thanks. This isn't required anymore. I'll have a look where it's used and remove it. We can directly use `vm.LogCfg`

fjl commented on tests/util.go:
```diff
@@ -171,9 +179,12 @@ func NewEnvFromMap(state *state.StateDB, envValues map[string]string, exeValues
 	env.gasLimit = common.Big(envValues["currentGasLimit"])
 	env.Gas = new(big.Int)
 
+	env.evm = vm.New(env, nil)
```
Note that this change causes an unexpected regression in the VM benchmarks because initialising the VM now does more work:

```text
~/d/e/s/g/e/go-ethereum >> ~/develop/eth/bin/benchstat old.txt new.txt
name                  old time/op  new time/op  delta
StateCall1024-8       12.6µs ± 1%  19.7µs ± 1%  +55.89%  (p=0.000 n=20+20)
VmAckermann32Tests-8  8.10ms ± 1%  8.22ms ± 1%   +1.53%  (p=0.000 n=19+17)
VmFibonacci16Tests-8  48.5ms ± 1%  48.8ms ± 1%   +0.77%  (p=0.000 n=20+19)
```

obscuren replied:
I think that's acceptable considering the isolation of the jump table (I think this is where the time is now spend)

fjl replied:
I agree that nothing needs to be done here, just wanted to inform you about it ;)

obscuren replied:
Thanks :-)

obscuren commented:
@fjl PTAL `StructLog` is now consistent and logger interface is removed. Also cleaned up the `eth/api`.`VmLoggerOptions`

fjl commented:
:+1:

obscuren commented:
@ppratscher *pling*
